demos/greengrass_connectivity/aws_greengrass_discovery_demo.c

```diff
@@ -85,134 +142,347 @@ typedef struct
+
+    /* Use the metrics string as username to report the OS and MQTT client version
+     * metrics to AWS IoT. */
+    xConnectInfo.pUserName = AWS_IOT_METRICS_STRING;
+    xConnectInfo.userNameLength = AWS_IOT_METRICS_STRING_LENGTH;
+
+    /* Set MQTT keep-alive period. If the application does not send packets at an interval less than
+     * the keep-alive period, the MQTT library will send PINGREQ packets. */
+    xConnectInfo.keepAliveSeconds = ggdDEMO_KEEP_ALIVE_INTERVAL_SECONDS;
+
+    /* Send MQTT CONNECT packet to broker. LWT is not used in this demo, so it
+     * is passed as NULL. */
+    xResult = MQTT_Connect( pxMQTTContext,
+                            &xConnectInfo,
+                            NULL,
+                            CONNACK_RECV_TIMEOUT_MS,
+                            &xSessionPresent );
+
+    if( xResult != MQTTSuccess )
+    {
+        LogError( ( "Failed to establish MQTT connection: Server=%s, MQTTStatus=%s", pxHostAddressData->pcHostAddress, MQTT_Status_strerror( xResult ) ) );
+    }
+    else
+    {
+        /* Successfully established MQTT connection with the broker. */
+        LogInfo( ( "An MQTT connection is established with %s.", pxHostAddressData->pcHostAddress ) );
+        xStatus = pdPASS;
+    }
+
+    return xStatus;
+}
 
-static IotMqttError_t _mqttConnect( GGD_HostAddressData_t * pxHostAddressData,
-                                    const IotNetworkInterface_t * pNetworkInterface,
-                                    IotMqttConnection_t * pMqttConnection )
+static BaseType_t prvConnectToServerWithBackoffRetries( GGD_HostAddressData_t * pxHostAddressData,
+                                                        NetworkContext_t * pxNetworkContext,
+                                                        MQTTContext_t * pxMQTTContext )
 {
-    IotMqttError_t xMqttStatus = IOT_MQTT_STATUS_PENDING;
-    IotNetworkServerInfo_t xServerInfo = { 0 };
-    IotNetworkCredentials_t xCredentials = AWS_IOT_NETWORK_CREDENTIALS_AFR_INITIALIZER;
-    IotMqttNetworkInfo_t xNetworkInfo = IOT_MQTT_NETWORK_INFO_INITIALIZER;
-    IotMqttConnectInfo_t xMqttConnectInfo = IOT_MQTT_CONNECT_INFO_INITIALIZER;
-    uint32_t connectAttempt = 0;
-
-    /* Set the server certificate for a secured connection. Other credentials
-     * are set by the initializer. */
-    xCredentials.pRootCa = pxHostAddressData->pcCertificate;
-    xCredentials.rootCaSize = ( size_t ) pxHostAddressData->ulCertificateSize;
-    /* Disable SNI. */
-    xCredentials.disableSni = true;
-    /* ALPN is not needed. */
-    xCredentials.pAlpnProtos = NULL;
-
-    /* Set the server info. */
+    ServerInfo_t xServerInfo = { 0 };
+
+    SocketsConfig_t xSocketsConfig = { 0 };
+    TransportSocketStatus_t xNetworkStatus = TRANSPORT_SOCKET_STATUS_SUCCESS;
+    BackoffAlgorithmContext_t xReconnectParams;
+    BaseType_t xBackoffStatus = pdFALSE;
+    BaseType_t xMqttStatus = pdFALSE;
```

Comment:
Nit: If you're using these variables in a pass/fail scenario, initializing them as such would be more consistent
```suggestion
    BaseType_t xBackoffStatus = pdFAIL;
    BaseType_t xMqttStatus = pdFAIL;
```